### PR TITLE
Fix caching issues with Event permissions

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Follow up to https://github.com/civicrm/civicrm-core/pull/12424

Calling checkPermission() with an event ID and then with no ID (to get all permissions) only returns permission for the first event, meaning all subsequent events are returned as having no permission when they probably do.

Before
----------------------------------------
If `CRM_Event_BAO_Event::checkPermission()` is called with a single event ID the static cache is "populated" but actually only has a single event ID in.  If the same function is called again with no ID the cache is returned rather than building an array of all events with permissions.

After
----------------------------------------
Refactor `CRM_Event_BAO_Event::checkPermission()` into two separate functions as they return two different things - an array of permissions in the case of all events, and TRUE/FALSE in the case of a single event ID.
A separate cache is implemented for all event permissions and for a single event permission.
So if you:
1. Call `checkPermission()` with an event ID you will get a TRUE/FALSE response and that will be cached.
2. Call `checkPermission()` with no event ID you will be "redirected" to `getAllPermissions()` which will build it's own cache and return an array of all event permissions.

Technical Details
----------------------------------------
If `CRM_Event_BAO_Event::checkPermission()` is called with a single event ID the static cache is "populated" but actually only has a single event ID in.  If the same function is called again with no ID the cache is returned rather than building an array of all events with permissions.

Comments
----------------------------------------
@alifrumin Would you be able to review this one?  It may need some further work on the existing unit tests as well if you have chance to take a look?

### CODE REVIEW

**Comment by alifrumin** [comment]
Hey @mattwire, I can take  a look at this too, I was waiting for the build to finish for this one and pr https://github.com/civicrm/civicrm-core/pull/12768 so I could test on the test build on jenkins, but it looks like that might be awhile so maybe I will test locally.

**Comment by alifrumin** [comment]
I tested this lightly locally,

I logged in as a user with the permission "CiviEvent: access CiviEvent" and could create but not edit an event. 

I got a permissions error immediately after creating the event: 

error

File: CRM/Event/BAO/Event.php [hunk 1]
```diff
@@ -2063,31 +2063,94 @@ public static function checkRegistration($params) {
...
    *
-   * @return string
-   *   the permission that the user has (or null)
+   * @return bool|array
+   *   Whether the user has permission for this event (or if eventId=NULL an array of permissions)
+   * @throws \CiviCRM_API3_Exception
    */
   public static function checkPermission($eventId = NULL, $type = CRM_Core_Permission::VIEW) {
+    if (empty($eventId)) {
+      return self::getAllPermissions($type);
```
Comment (eileenmcnaughton) [inline CRM/Event/BAO/Event.php]: agree - ideally add a deprecation notice

File: CRM/Event/BAO/Event.php [hunk 2]
```diff
@@ -2063,31 +2063,94 @@ public static function checkRegistration($params) {
...
+          if (isset($eventDetails['created_id']) && $contactId == $eventDetail['created_id']) {
+            $createdEvents[$eventId] = $eventId;
+          }
+        }
+      }
+
+      // Note: for a multisite setup, a user with edit all events, can edit all events
+      // including those from other sites
+      if (($type == CRM_Core_Permission::EDIT) && CRM_Core_Permission::check('edit all events')) {
+        Civi::$statics[__CLASS__]['permission'][$eventId] = TRUE;
```
Comment (eileenmcnaughton) [inline CRM/Event/BAO/Event.php]: I think you need an action nuance in your keys? ie 

Civi::$statics[__CLASS__]['permission']['edit'][$eventId] = TRUE;
Civi::$statics[__CLASS__]['permission']['view'][$eventId] = TRUE;

(set both on edit perm, only one on view )

Comment (mattwire) [reply CRM/Event/BAO/Event.php]: No. It's one of the magical mystery functions that returns different things depending on what you feed it.

**Comment by mattwire** [comment]
@alifrumin Do you have time to test this again?  I've refactored it since you found the original error but it could really do with some more tests as it was a very messy function before.

**Comment by mattwire** [comment]
@spalmstr Are you able to test this please? It relates to #11716

**Comment by alifrumin** [comment]
Taking a look now

**Comment by alifrumin** [comment]
I tested this by:

1. logging into the demo site created by jenkins.
2. creating a new drupal user "event" with the role "authenticated user"
3. giving the "authenticated user" role the following permissions in civi:
  + CiviEvent: access CiviEvent
  + CiviCRM: access CiviCRM backend and API
  + CiviEvent: view event info -- I was prompted to when trying to view event info however according to the description of CiviEvent: access CiviEvent - "Create events, view all events, and view participant records (for visible contacts)" I should not need it.

4. Logging in as the user "event"
5. Attempting to create an event - I am able to see the form to create an event, however on save of the event, it seems the event is created but I do not have permission to edit it the event, see screenshot below. This is the same behavior as described on September 7th.

error

**Comment by mattwire** [comment]
@alifrumin Thanks.  I just found the same thing :-( and realised how fragile it all is.  So I've just written a pile of tests.  `testDeleteOwnEvent()` will fail currently... is that even possible? And, if so, what permissions do they need to do that?

**Comment by mattwire** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
I worked through this & it looks pretty good -some thoughts

1)  I'd like to see calling checkPermission deprecated - at least to the extent of a code comment saying that - I realise more deprecation in this PR might be scope creep.

2) I think you could get some efficiencies by setting 'view' keys when you set edit keys. 

3) I find the 2 statics key names kinda confusing 'permission' vs 'permissions' 

4) I think you could check the 'permissions' statics before doing a query - since it's like that 'all events a contact can access' would be loaded in one call & then 'can they access x event ' in a second one

5) I think the rule for delete should be that they have 'delete in CiviEvent' AND edit access to the event in question. Arguably they should be able to delete events they created without 'delete in CiviEvent' but I think I'd be more worried about accidentally opening up delete permissions than accidentally reducing them

e.g

```
        if (CRM_Core_Permission::check('delete in CiviEvent') && CRM_Event_BAO_Event::checkPermission($eventId, CRM_Core_Permission::EDIT)) {
          Civi::$statics[__CLASS__]['permission']['delete'][$eventId] = TRUE;
        }
```

(you'd need to change the tests slightly for this assumption)

**Comment by eileenmcnaughton** [comment]
@mattwire style error

**Comment by alifrumin** [comment]
I tested this and it looks good to me! 

There is an error that shows up on the events dashboard for a user with the permissions:

+ CiviEvent: access CiviEvent
+ CiviCRM: access CiviCRM backend and API
+ CiviEvent: view event info 

but it sounds like that might be on purpose.

Error text:

> "User deprecated function: Deprecated function CRM_Event_BAO_Event::checkPermission, use CRM_Event_BAO_Event::getAllPermissions. Array ( [civi.tag] => deprecated ) in CRM_Core_Error_Log->log() (line 69 of /home/jenkins/bknix-dfl/build/core-12769-pu8u/sites/all/modules/civicrm/CRM/Core/Error/Log.php)." 

userdeprecatedfunction

**Comment by seamuslee001** [comment]
Given Eileen's comments and Alice's test i believe this is good to merge

(*CiviCRM Review Template WORD-1.1*)



* ([`r-explain`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-explain)) Pass
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) Pass - Expands unit tests as well
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) Pass
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) Pass
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) Pass
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) Pass - Tested by Alice and Eileen
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) Pass
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) Pass

**Comment by eileenmcnaughton** [comment]
Thanks @seamuslee001  - @mattwire @alifrumin 

We should try to fix those places where it calls the now-deprecated version of the function

**Pull request merged**